marlowe/src/Language/Marlowe/Client.hs

```diff
@@ -105,9 +112,97 @@ data PartyAction
 
 type RoleOwners = AssocMap.Map Val.TokenName PubKeyHash
 
-marlowePlutusContract :: Contract () MarloweSchema MarloweError ()
+type MarloweContractState = [MarloweData]
+
+data ContractHistory
+        = None
+        | Created MarloweParams MarloweData
+        | Transition TransactionInput
+        | History MarloweParams MarloweData [TransactionInput]
+    deriving (Show,Generic)
+    deriving anyclass (FromJSON, ToJSON)
+
+instance Semigroup ContractHistory where
+    any <> None                                  = any
+    _ <> Created params md                       = History params md []
+    History params md inputs <> Transition input = History params md (inputs <> [input])
+    cur <> _                                     = cur
+
+instance Monoid ContractHistory where
+    mempty = None
+
+marloweFollowContract :: Contract ContractHistory MarloweFollowSchema MarloweError ()
+marloweFollowContract = do
+    params <- endpoint @"follow"
+    slot <- currentSlot
+    logDebug @String "Getting contract history"
+    history slot params
+    logDebug @String "Start following"
+    follow params (slot + 5)
+  where
+    follow params timeout = do
+        wr <- SM.waitForUpdateUntil (mkMarloweClient params) timeout
+        logDebug @String ("State " <> show (encode wr))
+        case wr of
+            ContractEnded -> do
+                logDebug @String "Contract Ended"
+                pure ()
+            Timeout _ -> do
+                logDebug @String "Contract Timeout"
+                follow params (timeout + 5)
+            WaitingResult marloweData -> do
+                logDebug @String (show marloweData)
+                follow params (timeout + 5)
+
+    history slot params = do
+        let client@StateMachineClient{scInstance} = mkMarloweClient params
+        let inst = validatorInstance scInstance
+        let address = Scripts.scriptAddress inst
+        AddressChangeResponse{acrTxns} <- addressChangeRequest
+                AddressChangeRequest
+                { acreqSlotRange = Interval.to slot
+                , acreqAddress = address
+                }
```

Comment:
Just a query. What is this `addressChangeRequest` for and why is necessary?

Reply:
This request asks ChainIndex to respond with all transactions that modify a given address within a given slot range. That's how we get a history of a contract, and also how we wait for updates.

Reply:
Ah, I see, thanks